### PR TITLE
Rate limited Scan Implementation.

### PR DESCRIPTION
**Pull request by anandswaminathan**
The rate limited scan tries to pace the scan speed using sleeps, and also handles provision throughput exceeded exception. In order to handle large item sizes, the application can use the page_size variable to pace the scan if needed.

@danielhochman

### CODE REVIEW

File: pynamodb/connection/base.py [hunk 1]
```diff
@@ -923,6 +924,121 @@ def get_item(self,
...
+            if not last_evaluated_key and consecutive_provision_throughput_exceeded_ex == 0:
+                return
+
+            current_time = time.time()
+
+            # elapsed_time_ms indicates the time taken in ms from the start of the
+            # throttled_scan call.
+            elapsed_time_ms = round((current_time - start_time) * 1000)
+            # elapsed_time_ms can be 0 or negative if there is a clock drift
+            if elapsed_time_ms < 1:
```
Comment (danielhochman) [inline pynamodb/connection/base.py]: could collapse this to single line on :1009 using max()

File: pynamodb/connection/base.py [hunk 2]
```diff
@@ -923,6 +924,121 @@ def get_item(self,
...
+                elapsed_time_ms = 1
+
+            if consecutive_provision_throughput_exceeded_ex == 0:
+                total_consumed_read_capacity += latest_scan_consumed_capacity
+                consumed_rate = total_consumed_read_capacity / elapsed_time_ms
```
Comment (danielhochman) [inline pynamodb/connection/base.py]: i would probably from __future__ import division at top so you don't need to worry about accidentally doing integer division

Comment (anandswaminathan) [inline pynamodb/connection/base.py]: Wow. Awesome this is. Never knew it existed.

File: pynamodb/connection/base.py [hunk 3]
```diff
@@ -923,6 +925,118 @@ def get_item(self,
...
+                          table_name,
+                          attributes_to_get=None,
+                          page_size=None,
+                          limit=None,
+                          conditional_operator=None,
+                          scan_filter=None,
+                          exclusive_start_key=None,
+                          segment=None,
+                          total_segments=None,
+                          model_cls=None,
```
Comment (danielhochman) [inline pynamodb/connection/base.py]: the model should never make its way into the connection

File: pynamodb/connection/base.py [hunk 4]
```diff
@@ -923,6 +925,109 @@ def get_item(self,
...
+                    last_evaluated_key = data.get(LAST_EVALUATED_KEY, None)
+                    consecutive_provision_throughput_exceeded_ex = 0
+                except ScanError as e:
+                    # Only retry if provision throughput is exceeded.
+                    if isinstance(e.cause, ClientError):
+                        code = e.cause.response['Error'].get('Code')
+                        if code == "ProvisionedThroughputExceededException":
+                            consecutive_provision_throughput_exceeded_ex += 1
+                            if consecutive_provision_throughput_exceeded_ex > max_consecutive_exceptions:
+                                raise # Max threshold reached
```
Comment (danielhochman) [inline pynamodb/connection/base.py]: pep8 says two space before inline comment

i would put them on line above the raise though

File: pynamodb/connection/table.py [hunk 5]
```diff
@@ -131,6 +131,37 @@ def get_item(self, hash_key, range_key=None, consistent_read=False, attributes_t
             consistent_read=consistent_read,
             attributes_to_get=attributes_to_get)
 
+    def rate_limited_scan(self,
+             attributes_to_get=None,
```
Comment (danielhochman) [inline pynamodb/connection/table.py]: either move self down a line or indent everything to same level as self

File: pynamodb/models.py [hunk 6]
```diff
@@ -585,6 +585,69 @@ def query(cls,
...
+            scan_filter=key_filter,
+            segment=segment,
+            total_segments=total_segments,
+            exclusive_start_key=last_evaluated_key,
+            time_out_seconds=time_out_seconds,
+            read_capacity_to_consume_per_second=read_capacity_to_consume_per_second,
+            max_sleep_between_retry=max_sleep_between_retry,
+            max_consecutive_exceptions=max_consecutive_exceptions,
+        )
+
```
Comment (danielhochman) [inline pynamodb/models.py]: extra blank line

File: pynamodb/connection/table.py (hunk 5 above)
Comment (anandswaminathan) [inline pynamodb/connection/table.py]: Hey. I am following the format that was present throughout the file. Making this change will make this method appear odd

**Pull request merged**